File: wasm/examples/go-encryption/encryption_test.go
package encryption

import (
	"bytes"
	"crypto/rand"
	"encoding/binary"
	"encoding/hex"
	"testing"
)

const simpleSchema = `
table SimpleMessage {
  public_text: string;
  secret_number: int (encrypted);
  secret_text: string (encrypted);
}
root_type SimpleMessage;
`

const sensorSchema = `
table SensorReading {
  device_id: string;
  timestamp: uint64;
  temperature: float (encrypted);
  raw_data: [ubyte] (encrypted);
  secret_message: string (encrypted);
}
root_type SensorReading;
`

// createSimpleFlatBuffer creates a minimal FlatBuffer for SimpleMessage
func createSimpleFlatBuffer() []byte {
	buf := make([]byte, 64)

	// Root offset points to table at offset 16
	binary.LittleEndian.PutUint32(buf[0:], 16)

	// VTable at offset 4
	binary.LittleEndian.PutUint16(buf[4:], 10)  // vtable size
	binary.LittleEndian.PutUint16(buf[6:], 12)  // table size
	binary.LittleEndian.PutUint16(buf[8:], 4)   // field 0 offset
	binary.LittleEndian.PutUint16(buf[10:], 8)  // field 1 offset
	binary.LittleEndian.PutUint16(buf[12:], 0)  // field 2 not present

	// Table at offset 16: soffset to vtable
	binary.LittleEndian.PutUint32(buf[16:], 12) // 16 - 4 = 12

	// Field 0 (string offset) at table+4 = offset 20
	binary.LittleEndian.PutUint32(buf[20:], 12) // points to string at 32

	// Field 1 (int32) at table+8 = offset 24
	binary.LittleEndian.PutUint32(buf[24:], 42) // secret_number = 42

	// String at offset 32
	binary.LittleEndian.PutUint32(buf[32:], 5) // length
	copy(buf[36:], "hello")

	return buf
}

func TestEncryptionContextFromBytes(t *testing.T) {
	key := make([]byte, 32)
	rand.Read(key)
	ctx := NewEncryptionContext(key)

	if !ctx.IsValid() {
		t.Error("context should be valid")
	}
}

func TestEncryptionContextFromHex(t *testing.T) {
	hexKey := "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
	ctx, err := NewEncryptionContextFromHex(hexKey)
	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
	if !ctx.IsValid() {
		t.Error("context should be valid")
	}
}

func TestEncryptionContextInvalidSize(t *testing.T) {
	key := make([]byte, 16) // Too short
	ctx := NewEncryptionContext(key)

	if ctx.IsValid() {
		t.Error("context should be invalid")
	}
}

func TestDeriveDifferentKeys(t *testing.T) {
	key := make([]byte, 32)
	rand.Read(key)
	ctx := NewEncryptionContext(key)

	key1 := ctx.DeriveFieldKey(1)
	key2 := ctx.DeriveFieldKey(2)

	if bytes.Equal(key1, key2) {
		t.Error("derived keys should differ")
	}
}

func TestEncryptDecryptBytes(t *testing.T) {
	key := make([]byte, 32)
	iv := make([]byte, 16)
	rand.Read(key)
	rand.Read(iv)

	original := []byte("Hello, World!")
	data := make([]byte, len(original))
	copy(data, original)

	EncryptBytes(data, key, iv)
	if bytes.Equal(data, original) {
		t.Error("data should change after encryption")
	}

	DecryptBytes(data, key, iv)
	if !bytes.Equal(data, original) {
		t.Error("data should match after decryption")
	}
}

func TestDifferentIVs(t *testing.T) {
	key := make([]byte, 32)
	iv1 := make([]byte, 16)
	iv2 := make([]byte, 16)
	rand.Read(key)
	rand.Read(iv1)
	rand.Read(iv2)

	plaintext := []byte("Test data")
	data1 := make([]byte, len(plaintext))
	data2 := make([]byte, len(plaintext))
	copy(data1, plaintext)
	copy(data2, plaintext)

	EncryptBytes(data1, key, iv1)
	EncryptBytes(data2, key, iv2)

	if bytes.Equal(data1, data2) {
		t.Error("different IVs should produce different ciphertext")
	}
}

func TestParseSimpleSchema(t *testing.T) {
	fields := ParseSchemaForEncryption(simpleSchema, "SimpleMessage")
	if len(fields) != 3 {
		t.Errorf("expected 3 fields, got %d", len(fields))
	}

	var publicField, secretNum, secretText *FieldInfo
	for i := range fields {
		switch fields[i].Name {
		case "public_text":
			publicField = &fields[i]
		case "secret_number":
			secretNum = &fields[i]
		case "secret_text":
			secretText = &fields[i]
		}
	}

	if publicField == nil || publicField.Encrypted {
		t.Error("public_text should not be encrypted")
	}
	if secretNum == nil || !secretNum.Encrypted {
		t.Error("secret_number should be encrypted")
	}
	if secretText == nil || !secretText.Encrypted {
		t.Error("secret_text should be encrypted")
	}
}

func TestParseVectorFields(t *testing.T) {
	fields := ParseSchemaForEncryption(sensorSchema, "SensorReading")

	var rawData *FieldInfo
	for i := range fields {
		if fields[i].Name == "raw_data" {
			rawData = &fields[i]
			break
		}
	}

	if rawData == nil {
		t.Fatal("raw_data field not found")
	}
	if !rawData.Encrypted {
		t.Error("raw_data should be encrypted")
	}
	if rawData.Type != "vector" {
		t.Errorf("raw_data should be vector type, got %s", rawData.Type)
	}
}

func TestEncryptBufferChangesData(t *testing.T) {
	buf := createSimpleFlatBuffer()
	key := make([]byte, 32)
	rand.Read(key)

	encrypted, err := EncryptBuffer(buf, simpleSchema, key, "SimpleMessage")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if bytes.Equal(encrypted, buf) {
		t.Error("buffer should change after encryption")
	}
}

func TestEncryptDecryptRoundtrip(t *testing.T) {
	buf := createSimpleFlatBuffer()
	key := make([]byte, 32)
	rand.Read(key)

	encrypted, err := EncryptBuffer(buf, simpleSchema, key, "SimpleMessage")
	if err != nil {
		t.Fatalf("encryption error: %v", err)
	}

	decrypted, err := DecryptBuffer(encrypted, simpleSchema, key, "SimpleMessage")
	if err != nil {
		t.Fatalf("decryption error: %v", err)
	}

	if !bytes.Equal(decrypted, buf) {
		t.Error("buffer should match after decrypt")
	}
}

func TestDifferentKeysDifferentCiphertext(t *testing.T) {
	buf := createSimpleFlatBuffer()
	key1 := make([]byte, 32)
	key2 := make([]byte, 32)
	rand.Read(key1)
	rand.Read(key2)

	encrypted1, _ := EncryptBuffer(buf, simpleSchema, key1, "SimpleMessage")
	encrypted2, _ := EncryptBuffer(buf, simpleSchema, key2, "SimpleMessage")

	if bytes.Equal(encrypted1, encrypted2) {
		t.Error("different keys should produce different ciphertext")
	}
}

func TestWrongKeyWrongResult(t *testing.T) {
	buf := createSimpleFlatBuffer()
	key1 := make([]byte, 32)
	key2 := make([]byte, 32)
	rand.Read(key1)
	rand.Read(key2)

	encrypted, _ := EncryptBuffer(buf, simpleSchema, key1, "SimpleMessage")
	wrongDecrypt, _ := DecryptBuffer(encrypted, simpleSchema, key2, "SimpleMessage")

	if bytes.Equal(wrongDecrypt, buf) {
		t.Error("wrong key should not decrypt correctly")
	}
}

func TestContextReuse(t *testing.T) {
	key := make([]byte, 32)
	rand.Read(key)
	ctx := NewEncryptionContext(key)

	buf1 := createSimpleFlatBuffer()
	buf2 := createSimpleFlatBuffer()

	enc1, _ := EncryptBufferWithContext(buf1, simpleSchema, ctx, "SimpleMessage")
	enc2, _ := EncryptBufferWithContext(buf2, simpleSchema, ctx, "SimpleMessage")

	dec1, _ := DecryptBufferWithContext(enc1, simpleSchema, ctx, "SimpleMessage")
	dec2, _ := DecryptBufferWithContext(enc2, simpleSchema, ctx, "SimpleMessage")

	if !bytes.Equal(dec1, buf1) {
		t.Error("first buffer should decrypt correctly")
	}
	if !bytes.Equal(dec2, buf2) {
		t.Error("second buffer should decrypt correctly")
	}
}

func TestKeyDerivationConsistency(t *testing.T) {
	key, _ := hex.DecodeString("0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef")

	ctx1 := NewEncryptionContext(key)
	ctx2 := NewEncryptionContext(key)

	derived1 := ctx1.DeriveFieldKey(5)
	derived2 := ctx2.DeriveFieldKey(5)

	if !bytes.Equal(derived1, derived2) {
		t.Error("key derivation should be deterministic")
	}
}

func TestEncryptionIsDeterministic(t *testing.T) {
	key := make([]byte, 32)
	rand.Read(key)
	ctx := NewEncryptionContext(key)

	buf := createSimpleFlatBuffer()

	enc1, _ := EncryptBufferWithContext(buf, simpleSchema, ctx, "SimpleMessage")
	enc2, _ := EncryptBufferWithContext(buf, simpleSchema, ctx, "SimpleMessage")

	if !bytes.Equal(enc1, enc2) {
		t.Error("encryption should be deterministic")
	}
}

func TestHexKeySameAsBytes(t *testing.T) {
	keyBytes, _ := hex.DecodeString("0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef")
	hexKey := "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"

	ctxBytes := NewEncryptionContext(keyBytes)
	ctxHex, _ := NewEncryptionContextFromHex(hexKey)

	buf := createSimpleFlatBuffer()

	encBytes, _ := EncryptBufferWithContext(buf, simpleSchema, ctxBytes, "SimpleMessage")
	encHex, _ := EncryptBufferWithContext(buf, simpleSchema, ctxHex, "SimpleMessage")

	if !bytes.Equal(encBytes, encHex) {
		t.Error("hex and bytes keys should produce same result")
	}
}

func TestInvalidKeyReturnsError(t *testing.T) {
	buf := createSimpleFlatBuffer()
	key := make([]byte, 16) // Invalid size

	_, err := EncryptBuffer(buf, simpleSchema, key, "SimpleMessage")
	if err == nil {
		t.Error("expected error for invalid key size")
	}
}

File: wasm/examples/deno-encryption/encryption_test.ts
/**
 * Deno Encryption Integration Tests
 *
 * Run with: deno test --allow-read
 */

import {
  assertEquals,
  assertNotEquals,
  assertThrows,
} from "https://deno.land/std@0.208.0/assert/mod.ts";

import {
  EncryptionContext,
  encryptBytes,
  decryptBytes,
  encryptBuffer,
  decryptBuffer,
  parseSchemaForEncryption,
} from "./encryption.ts";

const SIMPLE_SCHEMA = `
table SimpleMessage {
  public_text: string;
  secret_number: int (encrypted);
  secret_text: string (encrypted);
}
root_type SimpleMessage;
`;

const SENSOR_SCHEMA = `
table SensorReading {
  device_id: string;
  timestamp: uint64;
  temperature: float (encrypted);
  raw_data: [ubyte] (encrypted);
  secret_message: string (encrypted);
}
root_type SensorReading;
`;

function createSimpleFlatBuffer(): Uint8Array {
  const buf = new Uint8Array(64);
  const view = new DataView(buf.buffer);

  // Root offset points to table at offset 16
  view.setUint32(0, 16, true);

  // VTable at offset 4
  view.setUint16(4, 10, true); // vtable size
  view.setUint16(6, 12, true); // table size
  view.setUint16(8, 4, true); // field 0 offset
  view.setUint16(10, 8, true); // field 1 offset
  view.setUint16(12, 0, true); // field 2 not present

  // Table at offset 16: soffset to vtable
  view.setUint32(16, 12, true); // 16 - 4 = 12

  // Field 0 (string offset) at table+4 = offset 20
  view.setUint32(20, 12, true); // points to string at 32

  // Field 1 (int32) at table+8 = offset 24
  view.setUint32(24, 42, true); // secret_number = 42

  // String at offset 32
  view.setUint32(32, 5, true); // length
  const hello = new TextEncoder().encode("hello");
  buf.set(hello, 36);

  return buf;
}

// Test 1: EncryptionContext
Deno.test("EncryptionContext - creates valid context from bytes", () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const ctx = new EncryptionContext(key);
  assertEquals(ctx.isValid(), true);
});

Deno.test("EncryptionContext - creates valid context from hex", () => {
  const hexKey = "0123456789abcdef".repeat(4);
  const ctx = new EncryptionContext(hexKey);
  assertEquals(ctx.isValid(), true);
});

Deno.test("EncryptionContext - rejects invalid key size", () => {
  const key = crypto.getRandomValues(new Uint8Array(16));
  const ctx = new EncryptionContext(key);
  assertEquals(ctx.isValid(), false);
});

Deno.test("EncryptionContext - derives different keys for different fields", () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const ctx = new EncryptionContext(key);
  const key1 = ctx.deriveFieldKey(1);
  const key2 = ctx.deriveFieldKey(2);
  assertNotEquals(Array.from(key1), Array.from(key2));
});

// Test 2: Low-level byte encryption
Deno.test("encryptBytes - encrypts and decrypts bytes", () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const iv = crypto.getRandomValues(new Uint8Array(16));
  const original = new TextEncoder().encode("Hello, World!");
  const data = new Uint8Array(original);

  encryptBytes(data, key, iv);
  assertNotEquals(Array.from(data), Array.from(original));

  decryptBytes(data, key, iv);
  assertEquals(Array.from(data), Array.from(original));
});

Deno.test("encryptBytes - different IVs produce different ciphertext", () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const iv1 = crypto.getRandomValues(new Uint8Array(16));
  const iv2 = crypto.getRandomValues(new Uint8Array(16));
  const plaintext = new TextEncoder().encode("Test data");

  const data1 = new Uint8Array(plaintext);
  const data2 = new Uint8Array(plaintext);

  encryptBytes(data1, key, iv1);
  encryptBytes(data2, key, iv2);

  assertNotEquals(Array.from(data1), Array.from(data2));
});

// Test 3: Schema parsing
Deno.test("parseSchemaForEncryption - parses simple schema", () => {
  const fields = parseSchemaForEncryption(SIMPLE_SCHEMA, "SimpleMessage");
  assertEquals(fields.length, 3);

  const publicField = fields.find((f) => f.name === "public_text");
  const secretNum = fields.find((f) => f.name === "secret_number");
  const secretText = fields.find((f) => f.name === "secret_text");

  assertEquals(publicField?.encrypted, false);
  assertEquals(secretNum?.encrypted, true);
  assertEquals(secretText?.encrypted, true);
});

Deno.test("parseSchemaForEncryption - parses vector fields", () => {
  const fields = parseSchemaForEncryption(SENSOR_SCHEMA, "SensorReading");
  const rawData = fields.find((f) => f.name === "raw_data");

  assertEquals(rawData?.encrypted, true);
  assertEquals(rawData?.type, "vector");
});

// Test 4: Buffer encryption
Deno.test("encryptBuffer - changes data", () => {
  const buf = createSimpleFlatBuffer();
  const original = new Uint8Array(buf);
  const key = crypto.getRandomValues(new Uint8Array(32));

  encryptBuffer(buf, SIMPLE_SCHEMA, key, "SimpleMessage");

  assertNotEquals(Array.from(buf), Array.from(original));
});

Deno.test("encryptBuffer - roundtrip", () => {
  const buf = createSimpleFlatBuffer();
  const original = new Uint8Array(buf);
  const key = crypto.getRandomValues(new Uint8Array(32));

  encryptBuffer(buf, SIMPLE_SCHEMA, key, "SimpleMessage");
  decryptBuffer(buf, SIMPLE_SCHEMA, key, "SimpleMessage");

  assertEquals(Array.from(buf), Array.from(original));
});

Deno.test("encryptBuffer - different keys produce different ciphertext", () => {
  const buf1 = createSimpleFlatBuffer();
  const buf2 = createSimpleFlatBuffer();
  const key1 = crypto.getRandomValues(new Uint8Array(32));
  const key2 = crypto.getRandomValues(new Uint8Array(32));

  encryptBuffer(buf1, SIMPLE_SCHEMA, key1, "SimpleMessage");
  encryptBuffer(buf2, SIMPLE_SCHEMA, key2, "SimpleMessage");

  assertNotEquals(Array.from(buf1), Array.from(buf2));
});

Deno.test("encryptBuffer - wrong key produces wrong result", () => {
  const buf = createSimpleFlatBuffer();
  const original = new Uint8Array(buf);
  const key1 = crypto.getRandomValues(new Uint8Array(32));
  const key2 = crypto.getRandomValues(new Uint8Array(32));

  encryptBuffer(buf, SIMPLE_SCHEMA, key1, "SimpleMessage");
  decryptBuffer(buf, SIMPLE_SCHEMA, key2, "SimpleMessage");

  assertNotEquals(Array.from(buf), Array.from(original));
});

// Test 5: Context reuse
Deno.test("EncryptionContext - can be reused", () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const ctx = new EncryptionContext(key);

  const buf1 = createSimpleFlatBuffer();
  const buf2 = createSimpleFlatBuffer();
  const orig1 = new Uint8Array(buf1);
  const orig2 = new Uint8Array(buf2);

  encryptBuffer(buf1, SIMPLE_SCHEMA, ctx, "SimpleMessage");
  encryptBuffer(buf2, SIMPLE_SCHEMA, ctx, "SimpleMessage");

  decryptBuffer(buf1, SIMPLE_SCHEMA, ctx, "SimpleMessage");
  decryptBuffer(buf2, SIMPLE_SCHEMA, ctx, "SimpleMessage");

  assertEquals(Array.from(buf1), Array.from(orig1));
  assertEquals(Array.from(buf2), Array.from(orig2));
});

// Test 6: Interoperability
Deno.test("key derivation is deterministic", () => {
  const key = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    key[i] = parseInt("0123456789abcdef"[i % 16], 16);
  }

  const ctx1 = new EncryptionContext(key);
  const ctx2 = new EncryptionContext(key);

  const derived1 = ctx1.deriveFieldKey(5);
  const derived2 = ctx2.deriveFieldKey(5);

  assertEquals(Array.from(derived1), Array.from(derived2));
});

Deno.test("encryption is deterministic", () => {
  const key = crypto.getRandomValues(new Uint8Array(32));
  const ctx = new EncryptionContext(key);

  const buf1 = createSimpleFlatBuffer();
  const buf2 = createSimpleFlatBuffer();

  encryptBuffer(buf1, SIMPLE_SCHEMA, ctx, "SimpleMessage");
  encryptBuffer(buf2, SIMPLE_SCHEMA, ctx, "SimpleMessage");

  assertEquals(Array.from(buf1), Array.from(buf2));
});

Deno.test("hex key same as bytes key", () => {
  const hexKey = "0123456789abcdef".repeat(4);
  const bytesKey = new Uint8Array(32);
  for (let i = 0; i < 64; i += 2) {
    bytesKey[i / 2] = parseInt(hexKey.substring(i, i + 2), 16);
  }

  const ctxHex = new EncryptionContext(hexKey);
  const ctxBytes = new EncryptionContext(bytesKey);

  const buf1 = createSimpleFlatBuffer();
  const buf2 = createSimpleFlatBuffer();

  encryptBuffer(buf1, SIMPLE_SCHEMA, ctxHex, "SimpleMessage");
  encryptBuffer(buf2, SIMPLE_SCHEMA, ctxBytes, "SimpleMessage");

  assertEquals(Array.from(buf1), Array.from(buf2));
});

// Test 7: Error handling
Deno.test("encryptBuffer - throws on invalid key", () => {
  const buf = createSimpleFlatBuffer();
  const key = crypto.getRandomValues(new Uint8Array(16)); // Invalid size

  assertThrows(
    () => encryptBuffer(buf, SIMPLE_SCHEMA, key, "SimpleMessage"),
    Error,
    "Invalid encryption key"
  );
});
